source/MaterialXCore/Document.cpp

```diff
@@ -826,10 +958,18 @@ void Document::upgradeVersion()
         minorVersion = 37;
     }
 
-    if (majorVersion == MATERIALX_MAJOR_VERSION &&
-        minorVersion == MATERIALX_MINOR_VERSION)
+    // Apply latest updates targetted for the next release version
+    if (applyLatestUpdates)
+    {
+        // Convert material Elements to Nodes
+        convertMaterialsToNodes(getDocument());
+        minorVersion = 38;
```

Comment:
With this latest round of changes, we're effectively enabling support for material nodes in 1.37 documents, as an optional, forward-looking feature that pipelines and tools can enable.  Within that new context, maybe it makes more sense to leave the minor version unchanged when performing this upgrade, so that we're not removing the ability to distinguish between 1.37 and 1.38 documents in the future.

Reply:
Isn't the opposite true? That if this is kept at 1.37 then we are unable to tell if the upgrade has occurred? i.e. it's not adding support but is a replacement. (I just found that it wasn't possible to actually remove material element class since it's required for upgrade purposes.).
BTW, The change for atan2() will be added in this forward-looking logic and will not be able to tell the difference unless we have a version bump.

When the next version comes out then this logic should be changed to be the same as for other version upgrades. (I need to add back in a check if this is 1.37 plus the feature enabling) before performing this. 

Currently we have dependents which are on 1.37 and we need to "downgrade" to convert material nodes back to material elements and hence check for a 1.38 version to revert back without first having to scan all the elements to find out if material nodes exist.

Reply:
This is a good question, and here is my current thinking:

For the upgrade from material elements to material nodes, the 1.37 codebase will support both options as equally valid (with this proposed changelist), so there is not a need to distinguish between them with versioning.  For the upgrade of the atan2 function signature, any single version of MaterialX is practically restricted to supporting one signature or the other, so that upcoming change will require a true version bump to distinguish between them.

When we ultimately update the latest MaterialX version from 1.37 to 1.38, we’ll remove support for material elements, and any such elements in 1.37 documents will be automatically upgraded to nodes, so there’s no situation (that I’m aware of) where our support for both options in the 1.37 codebase creates ambiguity.

Does that ring true to you, or are there cases I’m overlooking?

Reply:
Yes. Having the separation based on what is not backwards compatible (signature change) is a good minimum criteria for a version bump. I will remove any explicit version checking but keep "existence" testing (does a material node exist) to allow for proper conversion.